src/hotspot/share/services/nmtPreInit.cpp

```diff
@@ -37,51 +37,66 @@ static void* raw_malloc(size_t s)               { ALLOW_C_FUNCTION(::malloc, ret
-static void fail_oom(size_t size) {
-  vm_exit_out_of_memory(size, OOM_MALLOC_ERROR, "VM early initialization phase");
-}
-
-// --------- NMTPreInitAllocation --------------
 
-NMTPreInitAllocation* NMTPreInitAllocation::do_alloc(size_t payload_size) {
-  const size_t outer_size = sizeof(NMTPreInitAllocation) + payload_size;
-  guarantee(outer_size > payload_size, "Overflow");
-  void* p = raw_malloc(outer_size);
+static void* raw_checked_malloc(size_t s) {
+  void* p = raw_malloc(s);
   if (p == nullptr) {
-    fail_oom(outer_size);
+    vm_exit_out_of_memory(s, OOM_MALLOC_ERROR, "VM early initialization phase");
   }
-  NMTPreInitAllocation* a = new(p) NMTPreInitAllocation(payload_size);
-  return a;
+  return p;
 }
 
-NMTPreInitAllocation* NMTPreInitAllocation::do_reallocate(NMTPreInitAllocation* old, size_t new_payload_size) {
-  assert(old->next == nullptr, "unhang from map first");
-  // We just reallocate the old block, header and all.
-  const size_t new_outer_size = sizeof(NMTPreInitAllocation) + new_payload_size;
-  guarantee(new_outer_size > new_payload_size, "Overflow");
-  void* p = raw_realloc(old, new_outer_size);
+static void* raw_checked_realloc(void* old, size_t s) {
+  void* p = raw_realloc(old, s);
   if (p == nullptr) {
-    fail_oom(new_outer_size);
+    vm_exit_out_of_memory(s, OOM_MALLOC_ERROR, "VM early initialization phase");
   }
-  // re-stamp header with new size
-  NMTPreInitAllocation* a = new(p) NMTPreInitAllocation(new_payload_size);
+  return p;
+}
+
+// --------- NMTPreInitAllocation --------------
+
+void* NMTPreInitAllocation::operator new(size_t count) {
+  return raw_checked_malloc(count);
+}
+
+void NMTPreInitAllocation::operator delete(void* p) {
+  return raw_free(p);
+}
+
+NMTPreInitAllocation* NMTPreInitAllocation::do_alloc(size_t payload_size) {
+  void* payload = raw_checked_malloc(payload_size);
+  NMTPreInitAllocation* a = new NMTPreInitAllocation(payload_size, payload);
   return a;
 }
 
-void NMTPreInitAllocation::do_free(NMTPreInitAllocation* p) {
-  assert(p->next == nullptr, "unhang from map first");
-  raw_free(p);
+NMTPreInitAllocation* NMTPreInitAllocation::do_reallocate(NMTPreInitAllocation* a, size_t new_payload_size) {
+  assert(a->next == nullptr, "unhang from map first");
+  void* new_payload = raw_checked_realloc(a->payload, new_payload_size);
+  NMTPreInitAllocation* a2 = new NMTPreInitAllocation(new_payload_size, new_payload);
+  delete a;
+  return a2;
+}
+
+void NMTPreInitAllocation::do_free(NMTPreInitAllocation* a) {
+  assert(a->next == nullptr, "unhang from map first");
+  raw_free(a->payload);
+  delete a;
 }
 
 // --------- NMTPreInitAllocationTable --------------
 
+void* NMTPreInitAllocationTable::operator new(size_t count) {
+  return raw_checked_malloc(count);
+}
+
+void NMTPreInitAllocationTable::operator delete(void* p) {
+  return raw_free(p);
+}
+
 NMTPreInitAllocationTable::NMTPreInitAllocationTable() {
   ::memset(_entries, 0, sizeof(_entries));
```

Comment:
Hmm. Couldn't we do away with the array of pointers and just make it flat as an array of NMTPreInitAllocation? Then we could treat it as a map of malloc pointer to NMTPreInitAllocation? This would then allow us to just free the entire table, headers included, if NMT is off.

Reply:
It is an open-addressing hashmap, so I could only inline the first layer of nodes. An early version of NMTPreInit had a closed-addressing map to keep all nodes as a flat array, but that limited the total number of possible pre-init allocations, and I did not want that. If some crazy person wants to call the JVM with a million arguments I won't deny him that.

Making only the first layer as a flat array would be possible, and - seeing how the load factor is so small - get rid of the lion's share of node allocations. But at the cost of more complexity (need to rewrite all node iterations) and more costs if the table is allocated (+~128K). Since complexity was one of the reasons stated to remove NMTPreInit, I try to avoid complexity as much as possible.